WordPress/Classes/ViewRelated/Reader/ReaderReblogAction.swift

```diff
@@ -0,0 +1,127 @@
+import Foundation
+
+/// Encapsulates a command to reblog a post
+class ReaderReblogAction {
+
+    private let blogService: BlogService
+    private let presenter: ReblogPresenter
+
+    init(blogService: BlogService? = nil,
+         presenter: ReblogPresenter = ReblogPresenter()) {
+        self.presenter = presenter
+
+        // fallback for self.blogService
+        func makeBlogService() -> BlogService {
+            let context = ContextManager.sharedInstance().mainContext
+            return BlogService(managedObjectContext: context)
+        }
+        self.blogService = blogService ?? makeBlogService()
+    }
+
+    /// Executes the reblog action on the origin UIViewController
+    func execute(readerPost: ReaderPost, origin: UIViewController) {
+        presenter.presentReblog(blogs: blogService.blogsForAllAccounts(),
+                                readerPost: readerPost,
+                                origin: origin)
+    }
+}
+
+/// Presents the approptiate reblog scene, depending on the number of available sites
+class ReblogPresenter {
+    private let postService: PostService
+
+    init(postService: PostService? = nil) {
+
+        // fallback for self.postService
+        func makePostService() -> PostService {
+            let context = ContextManager.sharedInstance().mainContext
+            return PostService(managedObjectContext: context)
+        }
+        self.postService = postService ?? makePostService()
+    }
+
+    /// Presents the reblog screen(s)
+    func presentReblog(blogs: [Blog],
+                       readerPost: ReaderPost,
+                       origin: UIViewController) {
+
+        switch blogs.count {
+        case 0:
+            break
+        case 1:
+            let post = postService.createDraftPost(for: blogs[0])
```

Comment:
Maybe `blogs.first` might be better?

Reply:
I know, the direct index does not look great. On the other hand, first returns an optional, which we can safely force unwrap in this case, but force unwrapping does not great too :) Which one do you prefer?

Reply:
Well, `blogs[0]` assumes the array is not empty, and this will crash if it is. It may be an edge case, but I think I'd be safer to ensure there is actually a blog here.

Reply:
ok!